### PR TITLE
Feature/storage s3

### PR DESCRIPTION
**Pull request by OmerGronich**

### CODE REVIEW

**Comment by drorrose** [comment]
@davidmeirlevy @OmerGronich 
On my computer couldn't run the service due to error (not finding aws-sdk, even though locally installed and in package.json). Does it reproduce on any of your computers?

**Comment by OmerGronich** [comment]
@davidmeirlevy @drorrose @or-yam 
Renaming asssets is still WIP.

**Comment by OmerGronich** [comment]
@davidmeirlevy  
in gcs services, when you load files you return the type property in the return value. 
gcs has the metadata.kind built in to the api but could not find a way to do this with s3.

`
async function loadFiles(storage, identifier = '/') {
  const gcs = new Gcs(storage)
  const fullPath = path.join(storage.metadata.basePath || '/', identifier)

  let list
  try {
    await gcs.ready
    list = await gcs.list(fullPath)
  } catch (e) {
    throw new Error(e.message || 'failed to get list of assets from: ' + fullPath)
  } finally {
    // run on background
    // TODO: reuse storage connection
    gcs.destroy()
  }

  return list.map((asset) => {
    const fileIdentifier = path.join(identifier, asset.metadata.name)

    return {
      name: asset.metadata.name,
      identifier: fileIdentifier,
      updated: asset.metadata.updated,
      **type: getAssetType(asset.metadata)**,
      publicUrl: joinUrl(storage.metadata.publicUrl, fileIdentifier)
    }
  })
}
`

**Comment by davidmeirlevy** [comment]
@OmerGronich rach provider use a different approach to distinguish the file types, but as an assets-service, we should at least tell the user is this is an image, a folder, or another "asset".

**Comment by OmerGronich** [comment]
> @OmerGronich rach provider use a different approach to distinguish the file types, but as an assets-service, we should at least tell the user is this is an image, a folder, or another "asset".

Got it 👍
I will figure it out.

File: server/models/s3.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+const { getSecret } = require("../../helpers/secrets-management");
+
+class S3 {
+  constructor(storage) {
+    this.name = storage.name;
+    this.ready = getSecret(storage.tenant, storage.authentication).then(
+      (decrypted) => {
+        const { accessKeyId, secretAccessKey } = decrypted.value;
+
+        AWS.config.setPromisesDependency();
```
Comment (davidmeirlevy) [inline server/models/s3.js:11]: This service is stateless and can run multiple times.. maybe this line should run only once?

File: server/models/s3.js [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        .promise();
+
+
+    } catch (e) {
+      throw { message: 'could not rename asset from storage: ' + oldFile };
+    }
+  }
+
+  async destroy() {
+
```
Comment (davidmeirlevy) [inline server/models/s3.js:111]: Should destroy the this._client connection.

**Review by davidmeirlevy — CHANGES_REQUESTED** [review]

File: server/models/s3.js (hunk 1 above)
Comment (OmerGronich) [reply server/models/s3.js]: 👍 no problem

moved outside of class -> [commit 277cb0895d8c838429c6a97533fbbb663b723c0c](https://github.com/greenpress/assets-service/pull/17/commits/277cb0895d8c838429c6a97533fbbb663b723c0c)

File: server/models/s3.js (hunk 2 above)
Comment (OmerGronich) [reply server/models/s3.js:133]: What does destroying the connection mean? 

All this._client requests are stateless, I'm pretty sure there is no continuous connection to S3. 
Can I delete this function?

File: server/models/s3.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+      }));
+
+      const folders = listedObjects.CommonPrefixes.map(({ Prefix }) => ({
+        metadata: {
+          name: Prefix,
+          kind: ASSET_TYPES.DIRECTORY
+        }
+      }));
+
+      return [...files, ...folders];
```
Comment (davidmeirlevy) [inline server/models/s3.js:62]: it's common to put folders first.

File: server/models/s3.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+          Prefix: path,
+          Delimiter: "/"
+        })
+        .promise();
+
+      const headParams = {
+        Bucket: this.bucket.name
+      };
+
+      const files = await Promise.all(listedObjects.Contents.map(async content => {
```
Comment (davidmeirlevy) [inline server/models/s3.js:41]: Is there no faster way to load basic data for the files?
there are a lot of requests here.

maybe the next phase is to start with the caching mechanism here.

File: server/models/s3.js [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+      const listedObjects = await this._client.listObjectsV2(params).promise();
+
+      if (listedObjects.Contents.length === 0) return;
+
+      const deleteParams = {
+        Bucket: this.bucket.name,
+        Delete: { Objects: [] }
+      };
+
+      listedObjects.Contents.forEach(({ Key }) => {
```
Comment (davidmeirlevy) [inline server/models/s3.js:102]: why not to use map, like:
```
      const deleteParams = {
        Bucket: this.bucket.name,
        Delete: { Objects: listedObjects.Contents.map({Key} => ({Key})) }
      };
```

File: server/models/s3.js [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        deleteParams.Delete.Objects.push({ Key });
+      });
+
+      await this._client.deleteObjects(deleteParams).promise();
+
+      if (listedObjects.IsTruncated) await this.remove(file);
```
Comment (davidmeirlevy) [inline server/models/s3.js:108]: this looks a bit recursive. why is that?

File: server/routes/assets.js [hunk 7]
```diff
@@ -1,14 +1,20 @@
...
+  getStorageAssets,
+  removeStorageAsset,
+  verifyIdentifier,
+  uploadStorageAssets,
+  renameStorageAssets
+} = require('../controllers/assets');
 
-function empty (_, res) {
-  return res.status(200).json({ message: 'endpoint not yet exists' }).end()
+function empty(_, res) {
```
Comment (davidmeirlevy) [inline server/routes/assets.js:12]: you can remove this function we don't use it anymore.

**Review by davidmeirlevy — COMMENTED** [review]
minor changes left. this can be approved, but it can be even better with those changes. :)

File: server/models/s3.js (hunk 3 above)
Comment (OmerGronich) [reply server/models/s3.js:62]: done

File: server/models/s3.js (hunk 4 above)
Comment (OmerGronich) [reply server/models/s3.js:41]: No problem! The code is no longer making the extra request to s3 to get the object metadata. Now deriving the asset type  with our getAssetType function.

File: server/models/s3.js (hunk 5 above)
Comment (OmerGronich) [reply server/models/s3.js:102]: Cool :) 
Refactored👍

File: server/routes/assets.js (hunk 7 above)
Comment (OmerGronich) [reply server/routes/assets.js:12]: done

File: server/models/s3.js (hunk 6 above)
Comment (OmerGronich) [reply server/models/s3.js:94]: the isTruncated boolean is s3's pagination flag. True = there are still assets left to retrieve

**Review by davidmeirlevy — APPROVED** [review]

**Pull request merged**